### ISSUE TITLE
Add support for exporting commands for Environment

### ISSUE THREAD

**Issue by xenoterracide** [systemd/systemd#20406]
Ok, I know this one has been discussed before, and turned down.  A search of the issues here didn't allow me to quickly find it though. I'd like to bring it up again as I have a nuanced problem.

As a developer, I occasionally need multiple versions of language runtimes.

The easiest way to support this is tools such as `nvm` and `sdkman` (sadly there is no one tool to rule them all). These tools export environment variables when you run their setup scripts, or whatever, in the shell. Systemd has made it impossible to export these environment variables to the GUI so that things, like IDE's, can use them. It would be nice to allow systemd to execute a script via the environment "services", so that these things can be set up. Also I suspect that it would be good for systemd to be able to publish any changes to these environment variables vi some API, so that things could receive updates if for example you change which version of node you're using.

**Comment by poettering** [comment]
I don't grok what you are asking for. what is "exporting commands for Environment" supposed to mean?

**Comment by xenoterracide** [comment]
`nvm` for example exports environment variables, but there isn't a good way to have that run so that other UI programs can benefit from it, and then use an `nvm` NodeJS. This is because, from what I can tell, systemd only offers static environment variable export on login.

**Comment by poettering** [comment]
I have no idea what nvm is. And I have no idea what "static environment variable export on login" is supposed to mean. You are speaking in riddles. I really don't grok what you are trying to say

**Comment by xenoterracide** [comment]
dude nvm https://www.google.com/search?client=firefox-b-1-d&q=nvm it's the first link for me... https://github.com/nvm-sh/nvm give it a try, look at what it does when you use it, and how it works, it exports environment variables when you switching versions. currently all the systemd stuff supports is configuring environment variables in a way where they can't change, and it doesn't support exporting them via an executed shell script. sdkman works the same way.

**Comment by poettering** [comment]
I am sorry, but I am not going to read through the sources of some huge shell script to figure out what it does. I have no experience with "Node". Not a web guy.

Please explain in clear words what you want, in a way us low-level people can understand it, without referring to software we are unlikely to know. Otherwise I cannot help you.

**Comment by xenoterracide** [comment]
I was figuring you'd look at the README, and maybe install it and try seeing what it did, not so much look at the internals.

 There's a class of “applications” that allow you to easily manage multiple versions of a programming language. `nvm` for Node, `rbenv` for Ruby, `sdkman` for Java, I think it's `pyenv` for python, `perlbrew` for Perl, etc. Honestly, I'm a bit surprised this doesn't exist for gcc and lvvm (maybe it does). With these applications you source their shell script, and that “installs” it, then you select the language version you want, and it updated environment variables.
 
 for example
 
 ```
 ❯ nvm install 16                                                                                 # ~
Downloading and installing node v16.6.1...
Downloading https://nodejs.org/dist/v16.6.1/node-v16.6.1-linux-x64.tar.xz...
############################################################################################## 100.0%
Computing checksum with sha256sum
Checksums matched!
Now using node v16.6.1 (npm v7.20.3)
Creating default alias: default -> 16 (-> v16.6.1)
❯ nvm version                                                                                    # ~
v16.6.1
❯ nvm use 14                                                                                     # ~
Now using node v14.17.4 (npm v6.14.14)
 ❯ env | grep -i NVM                                                                              # ~
PATH=/home/xeno/.local/share/nvm/versions/node/v14.17.4/bin:/home/xeno/.local/share/sdkman/candidates
/java/current/bin:/home/xeno/.zinit/plugins/sdkman:/home/xeno/.zinit/snippets/cpanmin.us:/home/xeno/.
zinit/plugins/zdharma---zsh-diff-so-fancy/bin:/home/xeno/.zinit/plugins/pavoljuhas---smart-change-dir
ectory:/home/xeno/.local/bin:/home/xeno/.zinit/plugins/zdharma---zconvey/cmds:/home/xeno/.zinit/polar
is/bin:/usr/local/bin:/usr/bin:/var/lib/snapd/snap/bin:/usr/local/sbin:/usr/lib/jvm/default/bin:/usr/
bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
NVM_DIR=/home/xeno/.local/share/nvm
NVM_CD_FLAGS=-q
NVM_RC_VERSION=
NVM_BIN=/home/xeno/.local/share/nvm/versions/node/v14.17.4/bin
NVM_INC=/home/xeno/.local/share/nvm/versions/node/v14.17.4/include/node
❯ nvm use 16                                                                                     # ~
Now using node v16.6.1 (npm v7.20.3)
❯ env | grep -i NVM                                                                              # ~
PATH=/home/xeno/.local/share/nvm/versions/node/v16.6.1/bin:/home/xeno/.local/share/sdkman/candidates/java/current/bin:/home/xeno/.zinit/plugins/sdkman:/home/xeno/.zinit/snippets/cpanmin.us:/home/xeno/.zinit/plugins/zdharma---zsh-diff-so-fancy/bin:/home/xeno/.zinit/plugins/pavoljuhas---smart-change-directory:/home/xeno/.local/bin:/home/xeno/.zinit/plugins/zdharma---zconvey/cmds:/home/xeno/.zinit/polaris/bin:/usr/local/bin:/usr/bin:/var/lib/snapd/snap/bin:/usr/local/sbin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
NVM_DIR=/home/xeno/.local/share/nvm
NVM_CD_FLAGS=-q
NVM_RC_VERSION=
NVM_BIN=/home/xeno/.local/share/nvm/versions/node/v16.6.1/bin
NVM_INC=/home/xeno/.local/share/nvm/versions/node/v16.6.1/include/node
```

So this presents 2 problems IMO. There's not a good way to export this to the gui, and especially say, the Desktop Environment.

Currently, I could make an Environment file and hardcode these values, and for things that aren't in the DE, e.g. gnome shell stuff, regular applications, I can simply make my own `.desktop` file. That however could also mess with how these applications work in the first place (not sure, I'm pretty sure executing the init (not sys v) script again would overwrite it).

If systemd could execute this script in a way that was guaranteed to run prior to the DE then that wouldn't be a problem. However, there is another problem, there's still no way to communicate updates. I'm thinking systemd with dbus might be able to have an API and a program to update exported environment programs, that applications, and the DE could use. 

The problem with using ~/.config/autostart is that's some kind of order sensitive. I'm surprised that works at all (I haven't tried it), because IMO, applications, or the DE shell should not have their environment modified by another application launching like that. It might be fair to have a `.desktop` that can launch earlier than `~/.config/autostart` before the launch of the DE, that is expected to have its environment affect the DE.

I know in the case of the first that there was previously a no answer somewhere, but I'm not sure if this use case was considered, the argument was, I believe, over sourcing `~/.profile`. 

I hope this helps clarify.

**Comment by poettering** [comment]
Why doesn't https://www.freedesktop.org/software/systemd/man/systemd.environment-generator.html suffice for you?

**Comment by xenoterracide** [comment]
That might for the first problem (I will test), although something that's unclear to me, will those register in --user directories? Requiring system access is a no go in my opinion. I don't see in the documentation that it searches in user directories, so that might need improving.

It doesn't fix the second problem though, which is being able to dynamically update the environment.

**Comment by xenoterracide** [comment]
> environment generators exist both for the system and the user manager.
Then maybe that documentation should be updated to include those paths.

> You can issue "systemctl daemon-reload" any time you like, which will rerun generators and reload config

sysemd will use that to an extent, but applications don't have a good, standard, way to receive that those have been updated do they?

**Comment by poettering** [comment]
> Then maybe that documentation should be updated to include those paths.

Hmm? Have you looked at that man page? Second line of the Synposys already should make this clear?

> sysemd will use that to an extent, but applications don't have a good, standard, way to receive that those have been updated do they? and to reload their process env (if that's really even that possible, heh, though I know spring could do it to some extent)

UNIX environment variables are only passed down the process tree when processes are forked off. Each process then gets its own copy of the env block and can make changes to it the way it likes, but there's no further propagation from elsewhere after the fork() happened.

The semantics of env vars are not something systemd came up with, it's how UNIX has worked since basically its inception.

What you are looking for doesn't exist in the env var concept, and systemd can't do anything about it, it's not involved in how env vars are passed around between processes.

Hence, let's close this, I don't think there's anything we could or should do here. If you don't like how env vars work, please talk to kernel people, but I doubt they'll be enthusiastic about your idea of propagating env vars outside of fork()

**Comment by xenoterracide** [comment]
I was looking at [environment-generator](https://www.freedesktop.org/software/systemd/man/systemd.environment-generator.html), which has no `~` or `$HOME` references in the synopsis or otherwise. You're saying it exists, though.

```
/usr/lib/systemd/system-environment-generators/some-generator

/usr/lib/systemd/user-environment-generators/some-generator

/run/systemd/system-environment-generators/*
/etc/systemd/system-environment-generators/*
/usr/local/lib/systemd/system-environment-generators/*
/usr/lib/systemd/system-environment-generators/*

/run/systemd/user-environment-generators/*
/etc/systemd/user-environment-generators/*
/usr/local/lib/systemd/user-environment-generators/*
/usr/lib/systemd/user-environment-generators/*
```

In contrast to (environment.d)[https://www.freedesktop.org/software/systemd/man/environment.d.html] and some others, you'll notice in this that `~/.config/...` is the first entry.
```
~/.config/environment.d/*.conf

/etc/environment.d/*.conf

/run/environment.d/*.conf

/usr/lib/environment.d/*.conf

/etc/environment
```

The man page,  is the same as the website, which means it does not contain anything about using the home directory or where these files should be put. So if that's supposed to work, I believe the synopsis needs to be updated to add the correct path

note: `man systemd-environment-d-generator` looks like it has some formatting issues to me.

**Comment by TriMoon** [comment]
@xenoterracide ,
The concept of Enviroment variables is not similar to "scoped" variables that propagate their current value.
You should view them more like static parameters passed to functions that won't change inside the function's lifetime.

Hence always start an issue explaining what you want to accomplish as "end result", and after that you can provide suggestions how to get that result with "some needed changes".

**Comment by xenoterracide** [comment]
Heh, I'm aware, it was one of those bad ideas that seemed like a good idea at the time. 

Even if you were supporting the idea of updating them via a bus, as a way of 12 factor configuration. 

Which would mean that an application would have to do something like springs property sources to get that update. This wouldn't make the update an actual environment variable, rather a config key of the same name. Even if all that happened, most applications would probably need not insignificant changes internally too receive those updates.

If you're not familiar with springs property sources, essentially CLI, env, yaml, etc, each get essentially their own map/dict/struct internally, and are searched for config values  via a sort of lookup path. Even with this though, spring apps aren't usually prepared for those to update after initialization.

My problem, and maybe I didn't mention it, is needing to have an app like nvm, which exports env vars, to be able to communicate with other apps, running apps, in a standardized way, that those variables have changed. So when those apps are forking new processes they can do so with the updates, and/or update their own internal configuration. Say I update the version of my programming language in, a konsole shell via something like nvm, it could broadcast the change of env vars to the bus. My IDE could be listening for env changes, so when that happens, it updates its config, and the next time it forks the language, it will use whatever I switched to, and whatever other variables that changed.

I probably explained this terribly, and wasn't thinking too deeply about it. I probably am still doing that.

Seemed like this was a good idea, but probably isn't.

**Comment by xenoterracide** [comment]
So where in `~/.config` does an environment generator go? Seeing as how, that isn't documented, but you say that it does work

**Comment by xenoterracide** [comment]
ok, well, I gave it a go, even tried logging in on a console-getty vt as my user. As far as I can tell, these don't do anything.

```
❯ cat /usr/lib/systemd/user-environment-generators/60-flatpak
#!/bin/sh
export GIO_USE_VFS=local
exec flatpak --print-updated-env
❯ cat /usr/lib/systemd/user-environment-generators/asdf.sh
#!/bin/sh

. /opt/asdf-vm/asdf.sh
```

I did a grep for GIO there was nothing. in my shell environment zshrc causes node to be this.

```
which node
/home/xeno/.asdf/shims/node
```

but if I stop that by switching to bash, or moving the file

```
[xeno@manjaro ~]$ which node
/usr/bin/node
```

So the generators either don't run, or don't work. I also tried `/etc/systemd/user-environment-generators`. Care to tell me what I'm doing wrong, and how to put it in the `~/.config` directory?

**Comment by xenoterracide** [comment]
this on the other hand works, but, it doesn't do what I'm talking about

```
[xeno@manjaro ~]$ cat ~/.config/environment.d/asdf.conf 
PATH=$HOME/.asdf/shims:$PATH
```

**Comment by TriMoon** [comment]
> My problem, and maybe I didn't mention it, is needing to have an app like nvm, which exports env vars, to be able to communicate with other apps, running apps, in a standardized way, that those variables have changed. So when those apps are forking new processes they can do so with the updates, and/or update their own internal configuration. Say I update the version of my programming language in, a konsole shell via something like nvm, it could broadcast the change of env vars to the bus. My IDE could be listening for env changes, so when that happens, it updates its config, and the next time it forks the language, it will use whatever I switched to, and whatever other variables that changed.

InUnix you deal with many processes/users, every process has it's own environment variables.
All processes forked from a process share the same variable values downwards on a **once-per fork** basis.
So you can't "propagate" environment variable changes on a global level and neither pass changes made in the parent to the forked one...
For the purpose for "passing" they are read-only or passed-once however you want to view it.

So in short if you make changes that other programs need to act upon, you need to restart those programs after the changes...
